nrkno/sofie-core
# feat: Implement QuickLoop (SOFIE-2878)

## About the Contributor

This pull request is posted on behalf of the BBC.

## Type of Contribution

This is a: 

Feature


## Current Behavior

Currently Sofie allows looping of an entire Rundown Playlist. Whether the playlist should be looping can only be set by the Blueprints (usually done according to properties coming from the NRCS).

## New Behavior


### What it does:

- The _QuickLoop_ feature enables users to define a looping section within Rundown Playlists using Start and End markers.
- Automatically loops between a designated section.

### How to use it:
1. Set Markers:
   - Right-click on a part within the Rundown Playlist.
   - Choose "Set as QuickLoop Start" or "Set as QuickLoop End" from the context menu.

2. Activate QuickLoop:
   - QuickLoop mode will automatically trigger when any part between the Start and End markers is taken.
   - If configured, parts within the QuickLoop will progress automatically.
  
3. Modify the QuickLoop:
   - Set new markers (by context-clicking) to extend or shorten the QuickLoop section.
   - Execute adlibs without interrupting the QuickLoop.
   - Ensure logical marker placement to prevent errors.

4. Exit QuickLoop Mode:
   - Clear the markers using the context menu, or set a part outside of the QuickLoop as _Next_.
   - Playback continues seamlessly after exiting QuickLoop Mode.


Deviations from the [RFC](https://github.com/nrkno/sofie-core/issues/1078), and some more details:
- A Rundown Playlist with user-defined QuickLoop has similar indication in the Lobby and Rundown View as a looping Playlist.
- Setting a Start marker on a part after the End marker results in clearing the End marker, and vice versa.
- Clearing/exiting the QuickLoop by the use of a hotkey or AdLib Action is not implemented.
- The following settings are available in the Generic properties of a Studio:
  - _Enable QuickLoop_ (enables QuickLoop context menu options, but doesn't affect Playlist looping enabled by the NRCS/blueprints)
  - _Force Auto in a Loop_, with three available options:
    - _Disabled_ (default - consistent with existing Playlist loop behaviour)
    - _Enabled, but skipping parts with undefined or 0 duration_
    - _Enabled on all Parts, applying QuickLoop Fallback Part Duration if needed_
  - _QuickLoop Fallback Part Duration_
- The QuickLoop section is indicated as shown below
![image](https://github.com/nrkno/sofie-core/assets/9103996/1706ae7c-0262-4de6-a736-c1d8b5e1b0cf)
![image](https://github.com/nrkno/sofie-core/assets/9103996/aec7c640-ef6a-458f-a110-fb7c609f3c6d)
![image](https://github.com/nrkno/sofie-core/assets/9103996/2235265b-9da9-4396-befe-e4a7c97978a9)
## Testing Instructions

Setting and Clearing Markers:

1. Open Sofie GUI and navigate to a Rundown Playlist that does not have looping enabled by the NRCS.
2. Right-click on a Part (as long as it is not marked as invalid) and verify that you can set a Start marker.
3. Right-click on another Part  (as long as it is not marked as invalid) and set the End marker.
4. Ensure that setting a Start marker after the End marker will clear the previous End marker, and vice versa.
5. Verify that you can clear the Start and End markers from the context menu.
6. Test that setting the Start and End markers on the same Part creates a single-part loop.

Loop Mode Functionality:

1. Set the Start and End markers within a Rundown Playlist.
2. Take a Part within the defined loop section.
3. Verify that the Loop Mode is triggered and subsequent Parts progress automatically.
4. Check that Parts within the loop show the AUTO indicator in the GUI.
5. Ensure that when the end of the loop is reached, the first item in the loop becomes Next and an Auto Take occurs.
6. Test exiting the Loop Mode using the dedicated hotkey, AdLib Action, or context menu item.
7. Confirm that played adlibs are removed each time the loop starts over.
8. Verify behavior when a dynamic part is inserted while playing the End part of the loop.
9. Test for Parts with a duration of 0 seconds or undefined within the looping section and verify the default behavior or any custom settings applied.
10. Check that all Parts outside of the loop are dimmed but retain functionality.
12. Verify the indication showing that the Loop Mode is active.

Additional Scenarios:

1. Test setting Start/End markers on different Parts to modify the looping section in the Loop Mode.
2. Verify that setting a Part outside of the looping section as Next and taking it exits the Loop Mode.
3. Ensure that ingest operations altering the order of Segments and Parts or removing Parts containing markers stop the Loop Mode and delete the markers.
4. Test ingest operations hiding the Segment containing marked parts and verify Loop Mode behavior.


Testing Instructions for Studio Settings:

Enable QuickLoop:

1. Verify that enabling QuickLoop in the Studio settings enables corresponding context menu options.

Force Auto in a Loop:

1. Test each option for "Force Auto in a Loop" to ensure they behave as described:
  - Verify that the "Disabled" option maintains consistency with existing Playlist loop behavior, not applying auto-next.
  - Validate the behavior of the "Enabled, but skipping parts with undefined or 0 duration" option.
  - Confirm that the "Enabled on all Parts, applying QuickLoop Fallback Part Duration if needed" option functions correctly.

QuickLoop Fallback Part Duration:

1. Verify that the QuickLoop Fallback Part Duration setting is applied appropriately in conjunction with corresponding Force Auto in a Loop value.

Regression testing instructions for the existing Playlist Looping Feature:

1. Validate that existing Blueprints continue to function without requiring modifications for Playlist looping.
2. Verify that Playlist looping enabled by the NRCS/Blueprints functions as expected and is indicated in the GUI as it used to be.
3. Confirm that Start and End markers are correctly locked to the Playlist start and end and the user can not override them.

## Other Information



## Status
